In [36]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [37]:
train = pd.read_csv('ML_train_2class.csv')
test = pd.read_csv('ML_test_2class.csv')

In [38]:
train.head()

Text  Sentiment
0  watched 34 movie wondered got horrible review ...          0
1  have never understood appeal garbo always come...          0
2  nacho vigalondo is famous spain is kind bad sh...          0
3  eric idle robbie coltraine janet suzman should...          0
4  great storyline message joan plowright is supe...          1

In [39]:
trial_train = train.copy()
trial_train.head()

Text  Sentiment
0  watched 34 movie wondered got horrible review ...          0
1  have never understood appeal garbo always come...          0
2  nacho vigalondo is famous spain is kind bad sh...          0
3  eric idle robbie coltraine janet suzman should...          0
4  great storyline message joan plowright is supe...          1

In [40]:
class2_X_train = trial_train['Text']
class2_y_train = trial_train['Sentiment']

In [41]:
trial_test = test.copy()
trial_test.head()

Text  Sentiment
0  others have already commented decline director...          0
1  play is cleverly constructed begin porter rain...          1
2  gettin sick movie sound entertaining oneline s...          0
3  ruin will brief is great acting funny line att...          1
4  knew would gave rent laugh maybe mindless fun ...          0

In [42]:
class2_X_test = trial_test['Text']
class2_y_test = trial_test['Sentiment']

# Vectorization

BOW Vectorizer

In [43]:
bow_vectorizer = CountVectorizer()
bow_vectorizer.fit(class2_X_train)

class2_bow_X_train = bow_vectorizer.transform(class2_X_train)
class2_bow_X_test = bow_vectorizer.transform(class2_X_test)

TF-IDF

In [44]:
tfidf_vectorizer = TfidfVectorizer()
tfidf_vectorizer.fit(class2_X_train)

class2_tfidf_X_train = tfidf_vectorizer.transform(class2_X_train)
class2_tfidf_X_test = tfidf_vectorizer.transform(class2_X_test)

# Tuning Naive Bayes with BOW

In [45]:
param_grid = {
    'alpha': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
    'force_alpha': [True, False],
    'fit_prior': [True, False]
}

In [46]:
nb_model = MultinomialNB()
clf = GridSearchCV(nb_model, param_grid=param_grid, cv=5, verbose=4, scoring='accuracy')

In [47]:
best_clf = clf.fit(class2_bow_X_train, class2_y_train)

Fitting 5 folds for each of 44 candidates, totalling 220 fits
[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.579 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.582 total time=   0.1s
[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.582 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.577 total time=   0.1s
[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.582 total time=   0.1s
[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.788 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.778 total time=   0.1s
[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.779 total time=   0.1s
[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.783 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.787 total time=   0.1s
[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.579 total time=   0.1s
[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.582 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.582 total time=   0.1s
[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.577 total time=   0.1s
[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.582 total time=   0.1s
[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.788 total time=   0.1s
[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.778 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.779 total time=   0.1s
[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.783 total time=   0.1s
[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.787 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.859 total time=   0.1s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.853 total time=   0.1s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.855 total time=   0.1s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.857 total time=   0.1s
[CV 5/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.859 total time=   0.1s
[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.859 total time=   0.1s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.853 total time=   0.1s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.855 total time=   0.1s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.857 total time=   0.1s
[CV 5/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.859 total time=   0.1s
[CV 1/5] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.859 total time=   0.1s
[CV 

In [48]:
# best parameters
best_clf.best_params_

{'alpha': 1.0, 'fit_prior': True, 'force_alpha': True}

In [49]:
# highest accuracy score
best_clf.best_score_

0.859975

Function to help us evaluate model performance

In [50]:
def train_and_eval(model, trainX, trainY, testX, testY):

    # training the model
    fitted_model = model.fit(trainX, trainY)

    # getting predictions
    y_preds_train = fitted_model.predict(trainX)
    y_preds_test = fitted_model.predict(testX)

    # evaluating the model
    print()
    print(model)
    print(f"Train accuracy score : {accuracy_score(trainY, y_preds_train)}")
    print(f"Test accuracy score : {accuracy_score(testY, y_preds_test)}")
    print(classification_report(testY, y_preds_test))
    print('\n',40*'-')

In [51]:
best_nb = MultinomialNB(alpha = 1.0, fit_prior = False, force_alpha = True)

In [52]:
train_and_eval(best_nb, class2_bow_X_train, class2_y_train, class2_bow_X_test, class2_y_test)


MultinomialNB(fit_prior=False, force_alpha=True)
Train accuracy score : 0.916925
Test accuracy score : 0.8603860386038604
              precision    recall  f1-score   support

           0       0.85      0.88      0.86      5000
           1       0.87      0.84      0.86      4999

    accuracy                           0.86      9999
   macro avg       0.86      0.86      0.86      9999
weighted avg       0.86      0.86      0.86      9999


 ----------------------------------------


Best Naive Bayes model with BOW gave us 86.0% accuracy on test data

# Tuning Naive Bayes with TF-IDF

In [53]:
param_grid = {
    'alpha': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
    'force_alpha': [True, False],
    'fit_prior': [True, False]
}

In [54]:
nb_model = MultinomialNB()
clf = GridSearchCV(nb_model, param_grid=param_grid, cv=5, verbose=4, scoring='accuracy')

In [55]:
best_clf = clf.fit(class2_tfidf_X_train, class2_y_train)

Fitting 5 folds for each of 44 candidates, totalling 220 fits
[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.581 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.583 total time=   0.1s
[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.582 total time=   0.1s
[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.578 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.585 total time=   0.1s
[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.778 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.768 total time=   0.1s
[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.773 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.772 total time=   0.1s
[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.777 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.581 total time=   0.1s
[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.583 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.582 total time=   0.1s
[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.578 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.585 total time=   0.1s
[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.778 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.768 total time=   0.1s
[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.773 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.772 total time=   0.1s
[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.777 total time=   0.1s
[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.866 total time=   0.1s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.858 total time=   0.1s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.860 total time=   0.1s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.861 total time=   0.1s
[CV 5/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.865 total time=   0.1s
[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.866 total time=   0.1s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.858 total time=   0.1s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.860 total time=   0.1s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.861 total time=   0.1s
[C

In [56]:
# best parameters
best_clf.best_params_

{'alpha': 0.8, 'fit_prior': True, 'force_alpha': True}

In [57]:
# highest accuracy score
best_clf.best_score_

0.8670500000000001

In [58]:
best_nb = MultinomialNB(alpha = 0.8, fit_prior = True, force_alpha = True)

In [59]:
train_and_eval(best_nb, class2_tfidf_X_train, class2_y_train, class2_tfidf_X_test, class2_y_test)


MultinomialNB(alpha=0.8, force_alpha=True)
Train accuracy score : 0.921475
Test accuracy score : 0.8678867886788679
              precision    recall  f1-score   support

           0       0.86      0.88      0.87      5000
           1       0.88      0.86      0.87      4999

    accuracy                           0.87      9999
   macro avg       0.87      0.87      0.87      9999
weighted avg       0.87      0.87      0.87      9999


 ----------------------------------------


Best Multinomial Naive Bayes with TF-IDF gives us 86.8% accuracy on test data

# Tuning Naive Bayes and changing TF-IDF parameters

In [ ]:
# ngram_range=(1, 3): This tells the vectorizer to consider unigrams, bigrams, and trigrams
# min_df=2: This means an n-gram must appear in at least two documents to be considered. This helps in removing very rare n-grams that might not be useful for modeling.
# max_df=0.85: This means an n-gram appearing in more than 85% of the documents will be ignored, helping in filtering out too common n-grams.
tfidf_vectorizer = TfidfVectorizer(ngram_range=(1, 3), min_df=2, max_df=0.85)
tfidf_vectorizer.fit(class2_X_train)

class2_tfidf_X_train = tfidf_vectorizer.transform(class2_X_train)
class2_tfidf_X_test = tfidf_vectorizer.transform(class2_X_test)

In [28]:
param_grid = {
    'alpha': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
    'force_alpha': [True, False],
    'fit_prior': [True, False]
}

In [29]:
nb_model = MultinomialNB()
clf = GridSearchCV(nb_model, param_grid=param_grid, cv=5, verbose=4, scoring='accuracy')

In [30]:
best_clf = clf.fit(class2_tfidf_X_train, class2_y_train)

Fitting 5 folds for each of 44 candidates, totalling 220 fits


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.504 total time=   0.4s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.505 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.505 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.505 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.508 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.841 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.842 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.848 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 4/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.843 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 5/5] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.846 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.504 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.505 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.505 total time=   0.4s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.505 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.508 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.841 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.842 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.848 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 4/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.843 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 5/5] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.846 total time=   0.3s
[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.902 total time=   0.3s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.895 total time=   0.3s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.899 total time=   0.3s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.896 total time=   0.4s
[CV 5/5] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.902 total time=   0.5s
[CV 1/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.902 total time=   0.5s
[CV 2/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.895 total time=   0.5s
[CV 3/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.899 total time=   0.5s
[CV 4/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.896 total time=   0.5s
[CV 5/5] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.902 total time=   0.3s
[CV

In [31]:
# best parameters
best_clf.best_params_

{'alpha': 0.2, 'fit_prior': True, 'force_alpha': True}

In [32]:
# highest accuracy score
best_clf.best_score_

0.8991750000000002

In [33]:
best_nb = MultinomialNB(alpha = 0.2, fit_prior = True, force_alpha = True)

In [34]:
train_and_eval(best_nb, class2_tfidf_X_train, class2_y_train, class2_tfidf_X_test, class2_y_test)


MultinomialNB(alpha=0.2, force_alpha=True)
Train accuracy score : 0.988425
Test accuracy score : 0.8928892889288929
              precision    recall  f1-score   support

           0       0.90      0.88      0.89      5000
           1       0.88      0.90      0.89      4999

    accuracy                           0.89      9999
   macro avg       0.89      0.89      0.89      9999
weighted avg       0.89      0.89      0.89      9999


 ----------------------------------------


Naive Bayes while changing TF-IDF parameters gives us an accuracy of 89.2% on test data